### DISCUSSION TITLE
Code structure and naming conventions like Laravel and Filament

### DISCUSSION THREAD

**Discussion by adevade** [lunarphp/lunar#1216] [Q&A]
I believe that both the Laravel Framework and Filament packages uses a similar style for their code structure and naming conventions, and are very well structured and easy to understand and get into.

For example placing `interfaces` in a `Contracts` folder[^1], and `traits` in a `Concerns` folder[^2].
Lunar places a `xInterface.php` next to the default class implementation[^3], and `traits` in a `Traits` folder[^4]. And sometimes without the Interface suffix[^5].

*Looking in to it more, Laravel actually seems to place some of their interfaces like you do[^6], next to the default implementation. Though, most that are supposed to be used in userland are still placed in a top `Contracts` folder[^7]. (I think&hellip;)*

I'm also a big fan of Filament because of everything being typed. It really helps the IDE understand the codebase easily. 

**What are your thoughts about aligning the Lunar codebase with these conventions, to make it feel intuitive and integrated for newcomers?**

> [!NOTE]
> I haven't looked very much into the Lunar codebase &mdash; yet &mdash; so please don't take this as critique. I'm just curious about what your thoughs and plans are. 😊 

[^1]: [Filament placing `interfaces` in a `Contracts` folder](https://github.com/filamentphp/filament/tree/3.x/packages/forms/src/Contracts)
[^2]: [Filament placing `traits` in a `Concerns` folder](https://github.com/filamentphp/filament/tree/3.x/packages/forms/src/Concerns)
[^3]: [Lunar placing `xInterface.php` next to default class implementation](https://github.com/lunarphp/lunar/blob/0.5/packages/core/src/Base/AttributeManifestInterface.php)
[^4]: [Lunar placing `traits` in a `Traits` folder](https://github.com/lunarphp/lunar/tree/0.5/packages/core/src/Base/Traits)
[^5]: [Lunar naming TaxDriver `interface` without suffix](https://github.com/lunarphp/lunar/blob/0.5/packages/core/src/Base/TaxDriver.php)
[^6]: [Laravel `xInterface.php` next to default class implementation](https://github.com/laravel/framework/blob/10.x/src/Illuminate/Database/ConnectionInterface.php)
[^7]: [Laravel placing `interfaces` in a top `Contracts` folder](https://github.com/laravel/framework/tree/10.x/src/Illuminate/Contracts)

**Comment by glennjacobs** [comment]
Yeah, these standards have started to come about recently. They were less known when we started the journey of this project.

I'm not against the idea of tidying it up to keep it inline. We have always tried to follow the "Laravel way" of doing things.

**Reply by adevade** [reply]
Yeah, better to do potential large breaking changes now before a v1.0 release.

Some more questions (if you have time):

- How many people are you, working on the project? (except for external contributors)
- Have you considered utilizing GitHub Projects for roadmaps? [Like Filament](https://github.com/orgs/filamentphp/projects/2) has been doing for v3. It really gives contributors an easy overview.
- Any timeframe on the Filament powered Hub?

Thanks! 🍻

**Reply by glennjacobs** [reply]
We are an agency, but we have 2 people actively working on it pretty much full time.

We did use GitHub projects for a roadmap, but found we couldn't keep it current due to ever changing requirements. We will likely re-introduce when things are more predictable.

Filament - still quite a new project, I would estimate a couple of months for first release - but don't quote me on that! :-)

**Reply by adevade** [reply]
Sounds good man! Looking forward to following and hopefully contributing to this project! I'll at least submit Swedish translations once everything's a bit more stable. 🎉 

I've been looking for something customizable for Laravel that feels good to use, cause I need to replace WooCommerce at work. It really is a complete pain to work with once you have to do anything custom.
